crates/biome_service/src/scanner/watcher.rs

```diff
@@ -235,61 +198,131 @@ impl WorkspaceWatcher {
+    ) -> Result<(), WorkspaceError> {
+        for path in paths {
+            workspace.index_folder(&path)?;
+        }
 
-        if let Err(error) = workspace.resync_file_through_watcher(&path, scan_kind) {
-            // TODO: Improve error propagation.
-            warn!("Error resyncing file {path}: {error}");
+        Ok(())
+    }
+
+    /// Indexes open one or more files or folders.
+    ///
+    /// If you already know the paths are folders, use [`Self::index_folders()`]
+    /// instead.
+    fn index_paths(
+        workspace: &impl WorkspaceWatcherBridge,
+        paths: Vec<Utf8PathBuf>,
+    ) -> Result<(), WorkspaceError> {
+        for path in paths {
+            Self::index_path(workspace, &path)?;
         }
+
+        Ok(())
     }
 
-    #[inline]
-    fn scan_kind_for_path(&self, path: &Path) -> Option<&ScanKind> {
-        self.watched_folders.get(path).or_else(|| {
-            path.parent()
-                .and_then(|dir_path| self.watched_folders.get(dir_path))
-        })
+    /// Indexes an individual file or folder.
+    ///
+    /// If you already know the path is a folder, use [`Self::index_folder()`]
+    /// instead. If you know it is a file, you should directly call
+    /// [`WorkspaceWatcherBridge::index_file()`] instead.
+    fn index_path(
+        workspace: &impl WorkspaceWatcherBridge,
+        path: &Utf8Path,
+    ) -> Result<(), WorkspaceError> {
+        if workspace.fs().path_is_dir(path) {
+            workspace.index_folder(path)
+        } else {
+            let Some(project_key) = workspace.find_project_for_path(path) else {
+                return Ok(()); // file events outside our projects can be safely ignored.
+            };
+
+            workspace.index_file(project_key, path)
+        }
+    }
+
+    /// Unloads the given `paths` from the workspace index.
+    fn unload_files(
+        workspace: &impl WorkspaceWatcherBridge,
+        paths: Vec<Utf8PathBuf>,
+    ) -> Result<(), WorkspaceError> {
+        for path in paths {
+            workspace.unload_file(&path)?;
+        }
+
+        Ok(())
+    }
+
+    /// Unloads the given `paths` from the workspace index.
+    ///
+    /// If you already know the paths are files, use [`Self::unload_files()`]
+    /// instead.
+    fn unload_paths(
+        workspace: &impl WorkspaceWatcherBridge,
+        paths: Vec<Utf8PathBuf>,
+    ) -> Result<(), WorkspaceError> {
+        for path in &paths {
+            workspace.unload_path(path)?;
+        }
+
+        Ok(())
+    }
+
+    fn rename_path(
+        workspace: &impl WorkspaceWatcherBridge,
+        from: &Utf8Path,
+        to: &Utf8Path,
+    ) -> Result<(), WorkspaceError> {
+        workspace.unload_path(from)?;
+        Self::index_path(workspace, to)
+    }
+
+    /// Reopens an individual file if the watcher (still) has interest in it.
+    #[tracing::instrument(level = "debug", skip(workspace))]
+    fn resync_file(workspace: &impl WorkspaceWatcherBridge, path: &Utf8Path) {
```

Comment:
Nit: for consistency with WatcherInstruction
```suggestion
    fn reindex_file(workspace: &impl WorkspaceWatcherBridge, path: &Utf8Path) {
```